Add completion for struct keys

mveytsman opened the pull request:
I noticed that autocompletion doesn't work for keys inside a struct.

Before
<img width="977" alt="Screen Shot 2021-12-11 at 7 48 57 AM" src="https://user-images.githubusercontent.com/34720/145677151-1be5761d-d016-441a-bb14-19ec0bb0df2e.png">

After
<img width="959" alt="Screen Shot 2021-12-11 at 7 47 46 AM" src="https://user-images.githubusercontent.com/34720/145677153-6bdaa1ff-723e-4e35-9e96-5000f6b2f3ef.png">

This is pretty much the code from `IEx.Autocomplete`.

mveytsman commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -207,6 +207,48 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
+  defp match_struct_fields_or_local_or_var(hint, ctx) do
+    case expand_struct_fields(ctx) do
+      {:ok, struct, fields} ->
+        for field <- fields, do: {:map_field, field, nil}
```
Is it okay to use `:map_field` here or should I add a new `:struct_field` type?

josevalim replied:
I am not sure how it works for the Monaco editor, but IEx behaves like this:

Take a NaiveDateTime struct, it has both `month` and `minute`. If you type `m`, you want to show both `month` and `minute` as options. However, if you type `mo`, you want to type show only `month` as option and, if the user press tab, you want to autocomplete to `month: `. In order to encode the additional `: `, you need a different type than `map_field.

jonatanklosko replied:
Yeah, having a separate item here makes sense, especially that we need to add the default value. To avoid confusion with access syntax maybe something like `:in_struct_field` (alternatives welcome :see_no_evil:).

@mveytsman so you can add `{:in_struct_field, name(), default :: value()}` and then handle this new item in `Livebook.Intellisense`.

mveytsman replied:
My first pass was to rename `:map_field` to `:map_field_access` since it correctly handles both map and struct accesses, and call this `:struct_field`

Do you still prefer `:in_struct_field` @jonatanklosko ?

jonatanklosko replied:
Generally `:map_field_access` sounds good, but it doesn't fit to "identifier" that much. Similarly we have `:function` and not `:function_call`, so I'm leaning towards keeping `:map_field` as is :thinking:

mveytsman replied:
As an argument for it, `:map_field_access` is used in IEx, but I see that it doesn't fit identifier. I'll keep thinking on it as I refactor this to work right :)

mveytsman commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -207,6 +207,48 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
+        |> Enum.reduce(Map.from_struct(mod.__struct__), fn {key, _}, map ->
+          Map.delete(map, key)
+        end)
+        |> Map.keys()
```
We could be propagating the default value for struct fields from here -- is there value in surfacing them in the autocompletion somehow?

josevalim replied:
I think we can show the default value in the editor, so I would return it! @jonatanklosko ?

jonatanklosko replied:
Yeah, we can make the doc say "Default: `<inspected>`" or something like this.

josevalim commented on test/livebook/intellisense_test.exs:
```diff
@@ -1053,6 +1053,42 @@ defmodule Livebook.IntellisenseTest do
              ] = Intellisense.get_completion_items("struct.my", binding, env)
     end
 
+    test "completion for struct keys inside struct" do
```
Can we try autocompletion of an exception and make sure the `__exception__` field does not show up? In the Elixir code we reject anything starting with `_` if the hint is empty.

mveytsman replied:
I added a test for this, but it's succeeding because we ignore exceptions from autocompletion entirely in `has_struct?`: https://github.com/livebook-dev/livebook/blob/6784fa9b46898189461ffd7d9d17c0d1f8b4424a/lib/livebook/intellisense/identifier_matcher.ex#L201-L204

This means that we don't provide autocompletion for exception fields when trying to access them:

<img width="969" alt="Screen Shot 2021-12-11 at 5 18 14 PM" src="https://user-images.githubusercontent.com/34720/145693463-6afb3e3d-3a8e-4008-8941-b1940efa7663.png">

To be honest I don't understand why that check is there

mveytsman replied:
I think my last line wasn't clear: I think completing `message` in this screenshot above is more correct and would match IEx, but I don't want to remove that check in `has_struct?` because I don't understand what it's for.

josevalim replied:
@jonatanklosko is there a reason why we exclude exceptions?

jonatanklosko replied:
We match the IEx behaviour here:

![image](https://user-images.githubusercontent.com/17034772/145711189-a2b07a9f-e650-4e16-a601-8b1c53a03297.png)

```elixir
iex(2)> e = %ArgumentError{}
%ArgumentError{message: "argument error"}
iex(3)> e.
__exception__    __struct__       message 
```

The exclusion is for completing struct creation:

![image](https://user-images.githubusercontent.com/17034772/145711202-694cce74-e2c5-41ae-9349-446b1c95775d.png)

---

Note: in order to complete map values we need to be in a new cell, so that we have access to that value.

mveytsman replied:
Ah so I used has_sruct in my code which now causes us to not autocomplete any keys inside exception structs

I'll pull apart the two checks so we're correct here

josevalim commented on test/livebook/intellisense_test.exs:
```diff
@@ -1089,6 +1086,21 @@ defmodule Livebook.IntellisenseTest do
+               }
+             ] =
+               Intellisense.get_completion_items(
+                 "%Livebook.IntellisenseTest.MyStruct{my",
```
To be clear, you need to declare an exception, with no fields, and then try to autocomplete this:

```suggestion
                 "%Livebook.IntellisenseTest.MyException{",
```

mveytsman replied:
oops! I think my comment above still stands though

josevalim replied:
Yes, the way the test passes is good still. :)

mveytsman replied:
Fixed with `%ArgumentError{`

josevalim commented on test/livebook/intellisense_test.exs:
```diff
@@ -1053,6 +1053,54 @@ defmodule Livebook.IntellisenseTest do
+               %{
+                 label: "__exception__"
+               }
+             ] =
```
I think this assertion is not really strong because if it returns two elements, one of them being `__exception__`, the test will still fail. What do you think about something like this:

    completion = ...
    refute Enum.find(completion, & &1.label == "__exception__")

mveytsman replied:
doh!

```
      refute Enum.find(completions, &match?(%{label: "__exception__"}, &1))
```

josevalim replied:
Yes, your suggestion also works!

jonatanklosko commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -48,7 +48,7 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
   def completion_identifiers(hint, binding, env) do
     context = Code.Fragment.cursor_context(hint)
-    ctx = %{binding: binding, env: env, matcher: @prefix_matcher}
+    ctx = %{binding: binding, env: env, matcher: @prefix_matcher, hint: hint}
```
We build similar context in `locate_identifier`, please add the hint there for completeness :)

mveytsman replied:
Thanks @jonatanklosko! I caught this and made this a draft as I should do a bigger refactor.

VSCode doesn't do anything when I mouse over a struct field, and I will follow that pattern only using them in autocomplete.

jonatanklosko replied:
It's probably fine to show it actually and I probably already works since `Code.Fragment.surround_context` handles it, right? Maybe showing the field itself is not particularly useful, but if we show the default value in docs it could, @josevalim?

josevalim replied:
`surround_context` doesn't handle struct fields. At the moment, I don't think handling it is very interesting either, all we can show is the default. My suggestion is to not add/worry about this for now.

josevalim replied:
Oh, you already handled it. So that's fine. :)

jonatanklosko commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -207,6 +207,48 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
+          Map.delete(map, key)
+        end)
+        |> Map.keys()
+        |> Enum.map(&Atom.to_string/1)
```
Here we can keep fields as atoms, see `@type identifier_item` at the top.

mveytsman replied:
@jonatanklosko we're a bit inconsistent here, as you can see `:map_field` is returning a string as the field name.  I can make it consistent in this pr! https://github.com/livebook-dev/livebook/blob/main/lib/livebook/intellisense/identifier_matcher.ex#L241

jonatanklosko replied:
Oh wow, now I see we actually have strings for everything except function and module. It used to be strings, but then we needed the function name as atom, so I switched it to atom. Looks like for other types I updated the typespec and missed the actual items :0 Feel free to update all of those, but I can also do it if you prefer :)

mveytsman replied:
Oh I see that -- I think I'm going to get lost trying to do it all :)

Would you like me to back out https://github.com/livebook-dev/livebook/pull/793/commits/019c656b2c11010d24d1a61fefc448cf0426a512 so you can fix it all in one swoop?

jonatanklosko replied:
No worries, keep that commit and I will update the rest later :)

mveytsman commented:
Okay I think this is ready for a second pass @jonatanklosko @josevalim - I believe I addressed all your concerns.

Completion:
<img width="956" alt="Screen Shot 2021-12-11 at 8 10 03 PM" src="https://user-images.githubusercontent.com/34720/145696798-890986fb-a0e8-4cce-bcf1-b11ebc7cdd0d.png">

Completion expansion:
<img width="980" alt="Screen Shot 2021-12-11 at 8 10 15 PM" src="https://user-images.githubusercontent.com/34720/145696800-8bc41dc4-0ab0-4258-850c-a1bc14f28237.png">

Location/hover
<img width="985" alt="Screen Shot 2021-12-11 at 8 10 27 PM" src="https://user-images.githubusercontent.com/34720/145696802-0434cdc4-ab68-4b75-b154-7c620dd115c7.png">

jonatanklosko commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -21,6 +21,7 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
   @type identifier_item ::
           {:variable, name(), value()}
           | {:map_field, name(), value()}
+          | {:in_struct_field, struct_name :: display_name(), name(), default :: value()}
```
I'd keep the actual module here and convert to string as in #798 (feel free to copy the conversion and I will rebase that PR later)

```suggestion
          | {:in_struct_field, module(), name(), default :: value()}
```

jonatanklosko commented on lib/livebook/intellisense.ex:
```diff
@@ -138,11 +138,20 @@ defmodule Livebook.Intellisense do
+      label: "#{name}",
+      kind: :field,
+      detail: "#{struct_name} struct field",
+      documentation: "Default: #{inspect(default)}",
```
This supports markdown, so we can use an elixir code snippet for highlighting. You can bring `value_snippet` back from https://github.com/livebook-dev/livebook/pull/756/files, then maybe like:

```
"""
**Default**

#{value_snippet(default, @line_length)}
"""
```

jonatanklosko commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -207,6 +225,61 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
+            :locate ->
+              map
+
+            # When autocompleting we remove the keys that have already been filled
```
When we are in `:locate` then `Core.Fragment.container_cursor_to_quoted` would replace the field key with `__cursor__`, and discarding other fields is fine. Or is there a case where we need this distinction?

mveytsman replied:
Ah I see! In that case I can remove the type from ctx again.

mveytsman replied:
On second thought we already have matcher in the ctx, i think type does belong there.

mveytsman commented:
I've addressed this round of feedback!

**Pretty printed defaults:**
<img width="290" alt="Screen Shot 2021-12-12 at 10 59 34 AM" src="https://user-images.githubusercontent.com/34720/145719634-4060dab9-ce68-462b-8e6f-0bd59300f49b.png">

jonatanklosko commented on lib/livebook/intellisense.ex:
```diff
@@ -138,11 +138,20 @@ defmodule Livebook.Intellisense do
+      label: "#{name}",
+      kind: :field,
+      detail: "#{inspect(struct)} struct field",
+      documentation: "Default: #{inspect(default)}",
```
Let's use the same markdown snippet here as in details for consistency :)

jonatanklosko commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -195,18 +213,73 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
+          |> Map.reject(fn {key, _} ->
+            key
+            |> Atom.to_string()
+            |> String.match?(~r/__.*__/)
```
```suggestion
            |> String.match?(~r/^__.*__$/)
```

jonatanklosko commented:
Beautiful, two tiny comments from me!

josevalim commented on lib/livebook/intellisense/identifier_matcher.ex:
```diff
@@ -195,18 +213,73 @@ defmodule Livebook.Intellisense.IdentifierMatcher do
+          |> Map.reject(fn {key, _} ->
+            key
+            |> Atom.to_string()
+            |> String.match?(~r/__.*__/)
```
We should remove any key with underscore. That's the convention for anything private in Elixir. `__...__` is about reflection.

josevalim replied:
```suggestion
            |> String.starts_with?("_")
```

jonatanklosko reviewed (approved):
Fantastic, thanks a lot! :shipit:

mveytsman commented:
Thanks! I learned a lot doing this 😁
